### PR TITLE
Dsc19 96 dhis orchestrator

### PR DESCRIPTION
**Pull request by bradsawadye**

### CODE REVIEW

File: docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    "transforms.patcourse_asymp": "event.dataValues[]+",
+    "transforms.Comcond_any": "event.dataValues[]+",
+    "transforms.patcourse_dateonset": "event.dataValues[]+",
+    "transforms.expo_travel": "event.dataValues[]+"
+  },
+  "inputTransforms": {
+    "patinfo_ID": "{ 'attribute': constants.attributes.patinfo_ID, 'value': requestBody.identifier }",
+    "patinfo_ageonset": "{ 'attribute': constants.attributes.patinfo_ageonset, 'value': requestBody.observations[0].value.valueQuantity.value }",
+    "patinfo_sex": "{ 'attribute': constants.attributes.patinfo_sex, 'value': constants.sexMap[requestBody.observations[1].value.valueCodeableConcept.coding[].code] }",
+    "caseId": "{ 'attribute': constants.attributes.caseId, 'value': $substringAfter($string($random()), '0.')}",
```
Comment (MattyJ007) [inline docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json]: DHIS2 generates this field automatically.
I tested this via tracker capture. Not sure about via the API though

File: docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    },
+
+    "transforms.patcourse_asymp": "event.dataValues[]+",
+    "transforms.Comcond_any": "event.dataValues[]+",
+    "transforms.patcourse_dateonset": "event.dataValues[]+",
+    "transforms.expo_travel": "event.dataValues[]+"
+  },
+  "inputTransforms": {
+    "patinfo_ID": "{ 'attribute': constants.attributes.patinfo_ID, 'value': requestBody.identifier }",
+    "patinfo_ageonset": "{ 'attribute': constants.attributes.patinfo_ageonset, 'value': requestBody.observations[0].value.valueQuantity.value }",
```
Comment (MattyJ007) [inline docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json]: These fields don't map directly you'll need a bit of logic to check the code alongside the integer in this answer. If the code is `a` we use the integer. If the code is `d` or `m`, fill in 0 (the patient is days or months old)

File: docker/importer/volume/surveillance-mediator/endpoint-handler.json [hunk 3]
```diff
@@ -127,14 +127,24 @@
     ],
     "response": [
       {
-        "id": "resource-resolver",
+        "id": "resource-resolver-fhir",
         "config": {
           "method": "POST",
           "url": "http://localhost:3003/resource-resolver",
```
Comment (MattyJ007) [inline docker/importer/volume/surveillance-mediator/endpoint-handler.json]: Perhaps you could change this path as well for consistency

**Review by MattyJ007 — CHANGES_REQUESTED** [review]
Looking very good - nice one.
Just a few small suggestions

File: docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  },
+  "transformation": {
+    "input": "JSON",
+    "output": "JSON"
+  },
+  "inputValidation": {
+    "type": "object",
+    "properties": {
+      "requestBody": {
+        "type": "object"
```
Comment (MattyJ007) [inline docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json]: A bit more validation would have highlighted the data issue here

```suggestion
        "type": "object",
        "properties": {
          "data": {
            "type": "object",
            "properties": {
              "identifier": {
                "type": "string"
              },
              "observations": {
                "type": "array"
              }
            },
            "required": ["identifier", "observations"]
          }
        },
        "required": ["data"]
```

File: docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      }
+    },
+
+    "transforms.patcourse_asymp": "event.dataValues[]+",
+    "transforms.Comcond_any": "event.dataValues[]+",
+    "transforms.patcourse_dateonset": "event.dataValues[]+",
+    "transforms.expo_travel": "event.dataValues[]+"
+  },
+  "inputTransforms": {
+    "patinfo_ID": "{ 'attribute': constants.attributes.patinfo_ID, 'value': requestBody.identifier }",
```
Comment (MattyJ007) [inline docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json]: Incorrect nesting
```suggestion
    "patinfo_ID": "{ 'attribute': constants.attributes.patinfo_ID, 'value': requestBody.data.identifier }",
```

Same for the other fields using requestBody

File: docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    "patinfo_ageonset": "{ 'attribute': constants.attributes.patinfo_ageonset, 'value': requestBody.observations[0].value.valueQuantity.value }",
+    "patinfo_sex": "{ 'attribute': constants.attributes.patinfo_sex, 'value': constants.sexMap[requestBody.observations[1].value.valueCodeableConcept.coding[].code] }",
```
Comment (MattyJ007) [inline docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json]: These constants maps aren't resolving...

File: docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    "type": "object",
+    "properties": {
+      "requestBody": {
+        "type": "object"
+      }
+    },
+    "required": ["requestBody"]
+  },
+  "inputMapping": {
+    "requestBody.identifier": "teiIdentifier",
```
Comment (MattyJ007) [inline docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json]: @bradsawadye 
Nesting
```suggestion
    "requestBody.data.identifier": "teiIdentifier",
```

File: docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json (hunk 1 above)
Comment (bradsawadye) [reply docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json]: addressed

File: docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json (hunk 2 above)
Comment (bradsawadye) [reply docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json]: addressed

File: docker/importer/volume/surveillance-mediator/endpoint-handler.json (hunk 3 above)
Comment (bradsawadye) [reply docker/importer/volume/surveillance-mediator/endpoint-handler.json]: addresed

File: docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json (hunk 5 above)
Comment (bradsawadye) [reply docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json]: addressed

File: docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json (hunk 6 above)
Comment (bradsawadye) [reply docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json]: addressed

File: docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json (hunk 7 above)
Comment (bradsawadye) [reply docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json]: addressed

**Review by MattyJ007 — CHANGES_REQUESTED** [review]
@bradsawadye the mapping solution you went with here has a lot of redundancy (all the duplications of the DHIS2 UIDs) Please revert [this commit](https://github.com/jembi/who-covid19-surveillance-package/pull/12/commits/9eaca97db99a65559ce5ec20bf11c2054e3eea6d) and simplify the implementation with the `$lookup()` JSONata function

File: docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    },
+    "transforms.patcourse_dateonset": "event.dataValues[]+"
+  },
+  "inputTransforms": {
+    "patinfo_ID": "{ 'attribute': constants.attributes.patinfo_ID, 'value': requestBody.data.identifier }",
+    "patinfo_ageonset": "{ 'attribute': constants.attributes.patinfo_ageonset, 'value': requestBody.data.observations[0].value.valueQuantity.value }",
+
+    "incidentDate": "$now()",
+
+    "patcourse_dateonset": "{ 'dataElement': constants.dataElements.patcourse_dateonset, 'value': requestBody.data.observations[4].value.valueDateTime }"
```
Comment (MattyJ007) [inline docker/importer/volume/surveillance-mediator/endpoint-resource-resolver-dhis.json]: @bradsawadye Something like this has no code duplication

```suggestion
    "patinfo_sex": "{ 'attribute': constants.attributes.patinfo_sex, 'value': $lookup(constants.sexMap, requestBody.data.observations[1].value.valueCodeableConcept.coding[0].code) }",

    "incidentDate": "$now()",
    "patcourse_asymp": "{ 'dataElement': constants.dataElements.patcourse_asymp, 'value': $lookup(constants.yesNoUnknownMap, requestBody.data.observations[3].value.valueCodeableConcept.coding[0].code) }",
    "Comcond_any": "{ 'dataElement': constants.dataElements.Comcond_any, 'value': $lookup(constants.yesNoUnknownMap, requestBody.data.observations[5].value.valueCodeableConcept.coding[0].code) }",
    "patcourse_dateonset": "{ 'dataElement': constants.dataElements.patcourse_dateonset, 'value': requestBody.data.observations[4].value.valueDateTime }",
    "expo_travel": "{ 'dataElement': constants.dataElements.expo_travel, 'value': $lookup(constants.yesNoUnknownMap, requestBody.data.observations[6].value.valueCodeableConcept.coding[0].code) }"
```

**Review by MattyJ007 — COMMENTED** [review]
@bradsawadye remember to update the mapping mediator version

**Review by MattyJ007 — APPROVED** [review]
Perfecto 👍🏼

**Pull request merged**